# Modules
A group of functions, variables and classes saved to a file, which is nothing but module. Every Python file (.py) acts as a module.<br>A module is like a container or file where you can store code (functions, classes, variables) to keep things organized.<br>
They help keep code neat, reusable, and easy to share with others.<br>

##### Characteristics of Module

- `Organize your code` by breaking it into smaller, logical pieces.
- `Reuse code` by importing functions and variables from one module into another script.
- `Maintain code` by isolating and encapsulating functionality into separate files.

### Types of Modules
Python provides several kinds of modules:

#### 1. Built-in Modules
Python comes with several built-in modules such as `math`, `os`, `sys`, `random`, `time`, and many others. These are part of the Python Standard Library and are pre-installed with Python, which makes them ready to use without any additional installation.

#### 2. User-Defined/Custom Modules
These are Python files that you create yourself. Any Python file (.py) can be treated as a module, which means you can write your own modules to hold reusable code.

#### 3. Third-Party Modules
These modules are written by others and are not included in the standard Python library. You can install them using package managers like pip. Examples include numpy, pandas, and requests.

### Importing Modules
To use a module in Python, you need to import it using the import keyword. Python provides several ways to import modules:

#### Import the Entire Module
This imports the whole module and allows you to access its components using the dot (.) operator.

In [2]:
import math
print(math.pi)

3.141592653589793


#### Import Specific Functions or Variables from a Module
You can import specific items from a module, which allows you to access them directly without needing the dot operator.<br>
we can access multiple function seperated by comma `,`.

In [3]:
from math import sqrt, pi

print(sqrt(25))
print(pi)

5.0
3.141592653589793


#### Import All Functions or Variables
If you want to import everything from a module, you can use the asterisk `*`, but this is generally discouraged as it can make it difficult to track where specific functions or variables come from.

In [4]:
from math import *

print(sin(90))  # as it works but when working with number of modules this thing can be little bit confusing

0.8939966636005579


#### Import with an Alias
You can give an alias to a module or function while importing. This is especially useful when the module name is long, or if you want to avoid name conflicts.<br>
Aliasing can be done with `as` keyword.

In [5]:
import math as m

print(m.sqrt(9))

3.0


### Modules functions

#### dir()
returns list of all members of a module.<br>
It includes all variables, classes and functions.

In [7]:
import math

dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'cbrt',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'exp2',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

In [10]:
import time
from imp import reload
import math
time.sleep(3)
reload(math)
time.sleep(3)
reload(math)
print("This is test file") 

This is test file


## OS Module

The `os` module in Python provides a way to interact with the operating system and perform tasks like handling files, directories, and system variables. 

In [3]:
import os

### File and Directory Operations
#### os.getcwd()
This function returns the current working directory (the folder where your script is running or where you are located in the terminal).

In [4]:
current_directory = os.getcwd()
print(current_directory) 

C:\Users\ASUS\anaconda_projects\prog_ds\python


#### os.chdir(path)
This function changes the current working directory to the specified path.

In [11]:
# Changing the current directory to '/path/to/directory'
# while giving path as input use forward slash "/"
os.chdir('C:/Users/ASUS/anaconda_projects/prog_ds')

# Verifying the change
print(os.getcwd())

C:\Users\ASUS\anaconda_projects\prog_ds


#### os.listdir(path='.')
This function lists all files and directories in the specified path. If no path is provided, it lists the contents of the current working directory.

In [13]:
# Open a file descriptor for reading
fd = os.open('example.txt', os.O_RDONLY)

# Always close the file descriptor after usage
os.close(fd)

#### os.mkdir(path)
This function creates a new directory at the specified path.<br>
If the file is already exists then it gives FileExistsError.

In [18]:
# Create a new directory named 'new_folder' in current working directory
os.mkdir('new_folder')

# Create a new directory in a specific path
os.mkdir('C:/Users/ASUS/anaconda_projects/prog_ds/python/new_folder')

#### os.makedirs(path)
This function creates intermediate directories if they do not exist. It is similar to mkdir(), but it can create nested directories in one call.

In [20]:
# Create nested directories
os.makedirs('C:/Users/ASUS/anaconda_projects/prog_ds/python/new_folder/new_folder')

#### os.remove(path)
This function removes (deletes) the specified file.

In [22]:
# Removing a file
os.remove('C:/Users/ASUS/anaconda_projects/prog_ds/example.txt')

#### os.rmdir(path)
This function removes an empty directory.

In [24]:
# Removing an empty directory
os.rmdir('C:/Users/ASUS/anaconda_projects/prog_ds/python/new_folder/new_folder')

#### os.removedirs(path)
This function removes directories recursively, meaning it will remove all empty directories in the specified path.

In [25]:
# Removing nested empty directories
os.removedirs('C:/Users/ASUS/anaconda_projects/prog_ds/python/new_folder')

#### os.rename(src, dst)
This function renames a file or directory from src to dst.

In [32]:
# Rename a file
os.rename('example.txt', 'new_name.txt')

# Rename a directory
os.rename('new_folder', 'new_directory')

#### os.path.exists(path)
This function returns True if the specified path exists, otherwise it returns False.

In [27]:
# Check if a file or directory exists
if os.path.exists('some_file_or_directory'):
    print("Path exists")
else:
    print("Path does not exist")

Path does not exist


#### os.path.isfile(path)
This function checks if the specified path is a file and returns True if it is, otherwise False.

In [28]:
# Check if a specific path is a file
if os.path.isfile('some_file.txt'):
    print("It is a file")
else:
    print("It is not a file")

It is not a file


#### os.path.isdir(path)
This function checks if the specified path is a directory.

In [29]:
# Check if a specific path is a directory
if os.path.isdir('some_directory'):
    print("It is a directory")
else:
    print("It is not a directory")

It is not a directory


### File Descriptor Operations
File descriptors are low-level representations of open files.

#### os.open(path, flags)
This function opens a file and returns its file descriptor. The flags argument determines the mode in which the file is opened (e.g., read-only, write-only, etc.).

In [37]:
# Open a file descriptor for reading
fd = os.open('new_name.txt', os.O_RDONLY)

# Always close the file descriptor after usage
# os.close(fd)

#### os.close(fd)
This function closes the file descriptor.

In [38]:
# Close the file descriptor
os.close(fd)

#### os.read(fd, n)
This function reads n bytes from the file descriptor fd.


In [46]:
# Reading 100 bytes from the file descriptor

fd = os.open('new_name.txt', os.O_RDONLY)
data = os.read(fd, 100)
print(data)

b'This file is an example file for working with os module.'


#### os.write(fd, string)
This function writes the specified string (in bytes) to the file descriptor.

In [47]:
# Writing to a file descriptor (must be opened in write mode)
fd = os.open('new_name.txt', os.O_WRONLY)
os.write(fd, b"Hello, world!")
os.close(fd)

In [49]:
# writing in a file overrides it already existing content
fd = os.open('new_name.txt', os.O_RDONLY)
data = os.read(fd, 100)
print(data)
os.close(fd)

b'Hello, world!an example file for working with os module.'


## Math Module
The math module in Python provides a wide range of mathematical functions and constants. 

### Mathematical Constants
The math module provides several useful constants:

- `math.pi`: The mathematical constant π (pi) approximately equal to 3.14159.
- `math.e`: The mathematical constant e (Euler’s number) approximately equal to 2.71828.
- `math.tau`: The mathematical constant τ (tau), which is 2π or approximately 6.28318.
- `math.inf`: Represents infinity.
- `math.nan`: Represents "Not-a-Number" (NaN).

In [12]:
import math

print(math.pi) 
print(math.e)  
print(math.tau)
print(math.inf)
print(math.nan)

3.141592653589793
2.718281828459045
6.283185307179586
inf
nan


### Number-Theoretic and Representation Functions
#### math.ceil(x)
Returns the smallest integer greater than or equal to x.

In [14]:
print(math.ceil(4.3)) 
print(math.ceil(-4.7)) 

5
-4


#### math.floor(x)
Returns the largest integer less than or equal to x.

In [15]:
print(math.floor(4.8))
print(math.floor(-4.2))

4
-5


#### math.trunc(x)
Returns the truncated integer value of x (removes the decimal part).

In [16]:
print(math.trunc(4.9))
print(math.trunc(-4.9))

4
-4


#### math.copysign(x, y)
Returns a float with the magnitude (absolute value) of x but the sign of y.

In [17]:
print(math.copysign(3, -0.0))
print(math.copysign(-5, 10))

-3.0
5.0


#### math.fabs(x)
Returns the absolute value of x as a float.

In [18]:
print(math.fabs(-5))

5.0


#### math.factorial(x)
Returns the factorial of x, an integer greater than or equal to 0.

In [19]:
print(math.factorial(5))  # (5 * 4 * 3 * 2 * 1 = 120)

120


#### math.gcd(a, b)
Returns the greatest common divisor of integers a and b.

In [20]:
print(math.gcd(48, 180))

12


### Power and Logarithmic Functions
#### math.exp(x)
Returns e raised to the power of x (e^x).

In [21]:
print(math.exp(1))   # (e^1 = e)
print(math.exp(3))   # (e^3)

2.718281828459045
20.085536923187668


#### math.log(x)
Returns the natural logarithm of x (logarithm to base e) or to a specified base.

In [26]:
print(math.log(10))       # (ln(10))
print(math.log(8, 2))     # (log base 2 of 8 = 3)

2.302585092994046
3.0


#### math.log2(x)
Returns the base-2 logarithm of x.

In [27]:
print(math.log2(16))  # (2^4 = 16)

4.0


#### math.log10(x)
Returns the base-10 logarithm of x.

In [28]:
print(math.log10(100))  # (10^2 = 100)

2.0


####  math.pow(x, y)
Returns x raised to the power of y (x^y).

In [29]:
print(math.pow(2, 3))   # (2^3)

8.0


####  math.sqrt(x)
Returns the square root of x.

In [30]:
print(math.sqrt(16))

4.0


### Trigonometric Functions
#### math.sin(x)
Returns the sine of x, where x is in radians.

In [34]:
print(math.sin(math.pi/2)) 

1.0


#### math.cos(x)
Returns the cosine of x, where x is in radians.

In [35]:
print(math.cos(0)) 

1.0


#### math.tan(x)
Returns the tangent of x, where x is in radians.

In [36]:
print(math.tan(math.pi/4)) 

0.9999999999999999


#### math.asin(x)
Returns the arc sine (inverse sine) of x, in radians. The result is between -π/2 and π/2.

In [38]:
print(math.asin(1))   # (π/2)

1.5707963267948966


#### math.acos(x)
Returns the arc cosine (inverse cosine) of x, in radians. The result is between 0 and π.

In [39]:
print(math.acos(1))

0.0


#### math.atan(x)
Returns the arc tangent (inverse tangent) of x, in radians. The result is between -π/2 and π/2.



In [40]:
print(math.atan(1))  # (π/4)

0.7853981633974483


#### math.atan2(y, x)
Returns the arc tangent of y/x, in radians. Unlike math.atan(x), this function accounts for the sign of both arguments to determine the correct quadrant.

In [41]:
print(math.atan2(1, 1))  # (π/4)

0.7853981633974483


#### math.degrees(x)
Converts an angle from radians to degrees.

In [42]:
print(math.degrees(math.pi)) 

180.0


#### math.radians(x)
Converts an angle from degrees to radians.

In [43]:
print(math.radians(180))  # (π)

3.141592653589793


###  Hyperbolic Functions
#### math.sinh(x)
Returns the hyperbolic sine of x.

In [45]:
print(math.sinh(1))

1.1752011936438014


#### math.cosh(x)
Returns the hyperbolic cosine of x.

In [46]:
print(math.cosh(1))

1.5430806348152437


#### math.tanh(x)
Returns the hyperbolic tangent of x.

In [47]:
print(math.tanh(1))

0.7615941559557649


#### math.asinh(x)
Returns the inverse hyperbolic sine of x.

In [48]:
print(math.asinh(1))

0.881373587019543


#### math.acosh(x)
Returns the inverse hyperbolic cosine of x.

In [49]:
print(math.acosh(1))  

0.0


#### math.atanh(x)
Returns the inverse hyperbolic tangent of x.

In [50]:
print(math.atanh(0.5))

0.5493061443340549


###  Special Functions
#### math.factorial(x)
Returns the factorial of x. x must be a non-negative integer.

In [51]:
print(math.factorial(5)) 

120


#### math.gamma(x)
Returns the Gamma function at x. The Gamma function generalizes the factorial function to non-integer values.

In [33]:
print(math.gamma(5))  # (because Gamma(5) = 4!)

24.0


#### math.lgamma(x)
Returns the natural logarithm of the absolute

In [32]:
print(math.lgamma(5))

3.178053830347945


# iterators

In [ ]:
iter(range(10))   # iter method is used to create iterator object

In [7]:
r = list(range(11))
print(r)
type(r)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


list

In [8]:
new = iter(range(9))
new

In [9]:
next(new)

0

In [10]:
for i in range(10):
    print(i, end=' ')

0 1 2 3 4 5 6 7 8 9 

In [11]:
for value in [2, 4, 6, 8, 10]:
    # do some operation
    print(value + 1, end=' ')

3 5 7 9 11 

In [13]:
I = iter([12,21,32])
type(I)
print(next(I))

12


In [16]:
L = iter([4,4646,74,352,76426,354])
print(L)

## an infinite for loop

In [ ]:
N = 10 ** 12
for i in range(N):
    i = 1
print(N)
    

In [ ]:
L = [2, 4, 6, 8, 10]
for i in range(len(L)):
    print(i, L[i])

## stopiteration statement

In [ ]:
# this statement is used to end the iteration 

class MyNumbers:
  def __iter__(self):
    self.a = 1
    return self

  def __next__(self):
    if self.a <= 20:
      x = self.a
      self.a += 1
      return x
    else:
      raise StopIteration

myclass = MyNumbers()
myiter = iter(myclass)

for x in myiter:
  print(x)


# Itertools module

#### groupby ( )

In [ ]:
a = '545536364747488'

In [ ]:
# calculates keys for each element present in the iterable

from itertools import groupby

print(*[(len(list(g)),int(k)) for k, g in groupby(input())])

5274332333336666
(1, 5) (1, 2) (1, 7) (1, 4) (2, 3) (1, 2) (5, 3) (4, 6)


In [ ]:
print(list(b))

[('5', <itertools._grouper object at 0x7f81093eebe0>), ('4', <itertools._grouper object at 0x7f81093ee160>), ('5', <itertools._grouper object at 0x7f81093ee400>), ('3', <itertools._grouper object at 0x7f81093ee8b0>), ('6', <itertools._grouper object at 0x7f81093eed00>), ('3', <itertools._grouper object at 0x7f81093ee9a0>), ('6', <itertools._grouper object at 0x7f81093ee910>), ('4', <itertools._grouper object at 0x7f81093eee50>), ('7', <itertools._grouper object at 0x7f81093eee20>), ('4', <itertools._grouper object at 0x7f81093ee880>), ('7', <itertools._grouper object at 0x7f81093ee5e0>), ('4', <itertools._grouper object at 0x7f81093ee730>), ('8', <itertools._grouper object at 0x7f81093ee7f0>)]


## Infinite iterators

#### count ( )

In [ ]:
# this count function acts as an infinite range

from itertools import count 

for i in count():
    if i >= 10:
        break
    print(i, end=', ')

0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 

In [ ]:
for i in count(10, 10):
  if i == 200:
    break
  print(i)

10
20
30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190


####cycle ( )

In [ ]:
from itertools import cycle

In [ ]:
# make an iterator returning element from the iterable

for i in cycle('abcd'):
  print(i, end=' ')

a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d a b c d 

KeyboardInterrupt: ignored

#### repeat ( )

In [ ]:
# returns repetitions of elements for no. of times

from itertools import repeat

r = repeat('abcd', 5)
print(*r)

abcd abcd abcd abcd abcd


## Combinatoric iterators

#### product ( )

In [ ]:
# this returns cartesian product of input iterables

from itertools import product
p = product('ab', range(3))
print(*p)

('a', 0) ('a', 1) ('a', 2) ('b', 0) ('b', 1) ('b', 2)


In [ ]:
from itertools import product
p = product(range(5), range(3))
print(*p)


(0, 0) (0, 1) (0, 2) (1, 0) (1, 1) (1, 2) (2, 0) (2, 1) (2, 2) (3, 0) (3, 1) (3, 2) (4, 0) (4, 1) (4, 2)


In [ ]:
p = product('abcd', 'xy', 'mn')
print(*p)

('a', 'x', 'm') ('a', 'x', 'n') ('a', 'y', 'm') ('a', 'y', 'n') ('b', 'x', 'm') ('b', 'x', 'n') ('b', 'y', 'm') ('b', 'y', 'n') ('c', 'x', 'm') ('c', 'x', 'n') ('c', 'y', 'm') ('c', 'y', 'n') ('d', 'x', 'm') ('d', 'x', 'n') ('d', 'y', 'm') ('d', 'y', 'n')


In [ ]:
print(*product('abcd'))

('a',) ('b',) ('c',) ('d',)


####permutations ( )



In [ ]:
# returns successive r length permutations of elements in the iterable.

from itertools import permutations
p = permutations(range(3))
print(type(p))
print(p)
print(*p)

<class 'itertools.permutations'>
(0, 1, 2) (0, 2, 1) (1, 0, 2) (1, 2, 0) (2, 0, 1) (2, 1, 0)


In [1]:
from itertools import permutations
p = permutations(range(4),3)
print(*p)

(0, 1, 2) (0, 1, 3) (0, 2, 1) (0, 2, 3) (0, 3, 1) (0, 3, 2) (1, 0, 2) (1, 0, 3) (1, 2, 0) (1, 2, 3) (1, 3, 0) (1, 3, 2) (2, 0, 1) (2, 0, 3) (2, 1, 0) (2, 1, 3) (2, 3, 0) (2, 3, 1) (3, 0, 1) (3, 0, 2) (3, 1, 0) (3, 1, 2) (3, 2, 0) (3, 2, 1)


#### combinations ( )

In [2]:
# returns r length subsequences of elements from the input iterable
# the combinations tuples are emitted in lexicographic ordering according to the order of the input iterable.

from itertools import combinations
c = combinations(range(4), 3)
print(*c)

(0, 1, 2) (0, 1, 3) (0, 2, 3) (1, 2, 3)


In [3]:
from itertools import combinations
c = list(combinations(range(4), 2))
print(*c)

(0, 1) (0, 2) (0, 3) (1, 2) (1, 3) (2, 3)


In [4]:
print(c)
print(list(c))
print(len(list(c)))
print(iter)
print([i for i in c])

[(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]
[(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]
6
<built-in function iter>
[(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]


#### combinations_with_replacement ( )

In [ ]:
# return r length subsequences of elements from the input iterable allowing individual elements to be repeated more than once

from itertools import combinations_with_replacement

c = combinations_with_replacement(range(4), 3)
print(*c)

(0, 0, 0) (0, 0, 1) (0, 0, 2) (0, 0, 3) (0, 1, 1) (0, 1, 2) (0, 1, 3) (0, 2, 2) (0, 2, 3) (0, 3, 3) (1, 1, 1) (1, 1, 2) (1, 1, 3) (1, 2, 2) (1, 2, 3) (1, 3, 3) (2, 2, 2) (2, 2, 3) (2, 3, 3) (3, 3, 3)


In [ ]:
print(*(''.join(i) for i in (product('ABCD', repeat=2))))
print(*(''.join(i) for i in (permutations('ABCD', 2))))
print(*(''.join(i) for i in (combinations('ABCD', 2))))
print(*(''.join(i) for i in (combinations_with_replacement('ABCD', 2))))

AA AB AC AD BA BB BC BD CA CB CC CD DA DB DC DD
AB AC AD BA BC BD CA CB CD DA DB DC
AB AC AD BC BD CD
AA AB AC AD BB BC BD CC CD DD
